### ISSUE TITLE
no-nesting: .then() in the reject function

### ISSUE THREAD

**Issue by EvgenyOrekhov** [xjamundx/eslint-plugin-promise#58]
I have the following code:

```js
const Promise = require("bluebird");
const db = require("./db");

// Connect to DB and apply migrations
// or retry calling itself every second if can't connect
module.exports = function connect(params) {
    return db
        .connect(params)
        .then(
            () => db.applyMigrations(),
            () => Promise.delay(1000).then(() => connect(params))
        );
};
```

I get the "Avoid nesting promises" warning for line

```js
            () => Promise.delay(1000).then(() => connect(params))
```

I can't think of a way to rewrite this code without nesting.
I think the "no-nesting" rule should allow using `.then()` in the reject function.

**Comment by xjamundx** [comment]
Could you try this:

```js
   return db
        .connect(params)
        .then(() => db.applyMigrations())
        .catch(() => Promise.delay(1000))
        .then(() => connect(params))
```
If that won't work then yes maybe you'll have to add an exception to this line

**Comment by EvgenyOrekhov** [comment]
@xjamundx That won't work, that would reconnect even after a successfull connection, thus creating an infinite recursion. The only workaround besides `// eslint-disable-line` is to introduce a new named function:

```js
    function retry() {
        return Promise.delay(1000).then(() => connect(params));
    }

    return db
        .connect(params)
        .then(
            () => db.applyMigrations(),
            retry
        );
```

It's just not possible to flatten the chain as the rule suggests.
I think the rule should warn only on the cases that could be simplified.

**Comment by xjamundx** [comment]
> that would reconnect even after a successfull connection

Ahh yes, you're right of course.

> I think the rule should warn only on the cases that could be simplified.

It's a good idea, but I'm not sure how to write an eslint rule for that. It's very tricky to do. Any suggestions?

>  The only workaround besides `// eslint-disable-line` is to introduce a new named function:

This rule is a style rule, it's helpful sometimes it's sometimes not helpful. In this case it may not be helpful and so maybe disabling it is appropriate.

Here's an alternative approach that might work (though it's a little weird)

```js
function doStuff(delay = 0)
    return Promise.delay(delay)
       .then(db.connect(params))
      .then(() => db.ApplyMigrations())
      .catch(() => doStuff(1000))
}
```

**Comment by EvgenyOrekhov** [comment]
@xjamundx Sorry, I'm not familiar with the development of ESLint rules.

Placing a `.catch()` after `.then(() => db.applyMigrations())` would catch errors thrown by `db.applyMigrations()`, and I don't need that, I only need to retry when `db.connect()` fails.

**Comment by xjamundx** [comment]
If you have ideas of how I may be able to detect cases that can't be streamlined, I can try to implement them. Here's a video of how to write eslint rules. it's really fun!

https://vimeo.com/119903352

As far as my weird restructuring idea. Works the same with `then()`:


Same, principal with `then`'s 2nd arg though:

```js
function doStuff(delay = 0, retry = true) {
    return Promise.delay(delay)
       .then(db.connect(params))
       .then(
            () => db.ApplyMigrations())
            () => retry ? doStuff(1000, false) : Promise.reject('bad')
       )
}
```
